cni/pkg/taint/taintcontroller.go

```diff
@@ -0,0 +1,414 @@
+	errorHandler := func(obj interface{}, pod *v1.Pod, err error) {
+		if err == nil {
+			log.Debugf("Removing %s/%s from work queue", pod.Namespace, pod.Name)
+			tc.podWorkQueue.Forget(obj)
+		} else if tc.podWorkQueue.NumRequeues(obj) < 50 {
+			log.Errorf("Error: %s", err.Error())
+			log.Infof("Re-adding %s/%s to work queue", pod.Namespace, pod.Name)
+			tc.podWorkQueue.AddRateLimited(obj)
+		} else {
+			log.Infof("Requeue limit reached, removing %s/%s", pod.Namespace, pod.Name)
+			tc.podWorkQueue.Forget(obj)
+			runtime.HandleError(err)
+		}
+	}
+	obj, quit := tc.podWorkQueue.Get()
+	if quit {
+		// Exit permanently
+		return false
+	}
+	defer tc.podWorkQueue.Done(obj)
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		log.Errorf("Error decoding object, invalid type. Dropping.")
+		tc.podWorkQueue.Forget(obj)
+		// Short-circuit on this item, but return true to keep
+		// processing.
+		return true
+	}
+	//check for pod readiness
+	//if pod is ready, check all of critical labels in the corresponding node, if all of them are ready, untaint the node
+	//if pod is not ready or some critical pods are not ready, taint the node
+	if podutils.IsPodReady(pod) {
+		err := tc.processReadyPod(pod)
+		errorHandler(obj, pod, err)
+	} else {
+		err := tc.processUnReadyPod(pod)
+		errorHandler(obj, pod, err)
+	}
+	// Return true to let the loop process the next item
+	return true
+}
+
+//list all candidate pods given node name, namespace and selector using cached storage
+func (tc Controller) listCandidatePods(nodeName string, namespace string, selector string) []*v1.Pod {
+	if _, ok := tc.cachedPodsStore[namespace]; !ok {
+		return []*v1.Pod{}
+	}
+	if _, ok := tc.cachedPodsStore[namespace][selector]; !ok {
+		return []*v1.Pod{}
+	}
+	podList := make([]*v1.Pod, 0)
+	for _, item := range tc.cachedPodsStore[namespace][selector].List() {
+		pod, ok := item.(*v1.Pod)
+		if !ok {
+			continue
+		}
+		if pod.Spec.NodeName == nodeName && podutils.IsPodReady(pod) {
+			podList = append(podList, pod)
+		}
+	}
+	return podList
+}
+
+func (tc Controller) CheckNodeReadiness(node v1.Node) bool {
+	if tc.taintsetter.configs == nil || len(tc.taintsetter.configs) == 0 {
+		return true
+	}
+	for namespace, labelMap := range tc.cachedPodsStore {
+		for label := range labelMap {
+			if len(tc.listCandidatePods(node.GetName(), namespace, label)) == 0 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+//if pod is ready, check all related pod in the corresponding node, if any critical pod is unready, it should be tainted
+func (tc Controller) processReadyPod(pod *v1.Pod) error {
+	node, err := tc.getNodeByPod(pod)
+	if err != nil {
+		return fmt.Errorf("Fatal error Cannot get node by  %s in namespace %s : %s", pod.Name, pod.Namespace, err)
+	}
+	if GetNodeLatestReadiness(*node) && tc.CheckNodeReadiness(*node) {
+		err = tc.taintsetter.RemoveReadinessTaint(node)
+		if err != nil {
+			return fmt.Errorf("Fatal error Cannot remove node readiness taint: %s ", err.Error())
```

Comment:
"fatal error" is misleading here as generally fatal==exit the program

Reply:
removed all fatal error prefix